modules/cudaoptflow/src/pyrlk.cpp

```diff
@@ -51,7 +51,93 @@ Ptr<cv::cuda::SparsePyrLKOpticalFlow> cv::cuda::SparsePyrLKOpticalFlow::create(S
 
 Ptr<cv::cuda::DensePyrLKOpticalFlow> cv::cuda::DensePyrLKOpticalFlow::create(Size, int, int, bool) { throw_no_cuda(); return Ptr<DensePyrLKOpticalFlow>(); }
 
-#else /* !defined (HAVE_CUDA) */
+#else
+
+void cv::cuda::buildImagePyramid(const GpuMat& Img, std::vector<GpuMat>& Pyr,
+                                 int maxLevel, Stream& stream)
+{
+    Pyr.resize(maxLevel + 1);
+    int cn = Img.channels();
+
+    CV_Assert(cn == 1 || cn == 3 || cn == 4);
+
+    Pyr[0] = Img;
+    for (int level = 1; level <= maxLevel; ++level)
+    {
+        cuda::pyrDown(Pyr[level - 1], Pyr[level], stream);
+    }
+}
+
+void sparse(std::vector<GpuMat>& prevPyr, std::vector<GpuMat>& nextPyr, const GpuMat& prevPts, GpuMat& nextPts, GpuMat& status, GpuMat* err,
+            int winSize[2], int halfWinSize[2], int maxLevel, int& iters, bool useInitialFlow, Stream& stream)
```

Comment:
I think that the above API is having a lot of parameters. Would it be good if we can make `sparse` a friend of `PyrLKOpticalFlowBase` so that it can use it's private members and the need for passing them as parameters is eradicated? What do you say @alalek ? I will make changes to the rest of the calls to `sparse` after finalising this API.

Reply:
ping @alalek

Reply:
Hi, @alalek @asmorkalov Please let me know your views on this thread. Further updates depend on the decisions taken here. Thanks.

Reply:
Hi, @alalek Please let me know your views on the API. Thank you.